build/ios/mojom/objc_templates/private_interface_implementation.tmpl

```diff
@@ -67,33 +67,20 @@ base::BindOnce(callback)
   return self;
 }
 
-- (void)dealloc {
-  web::GetUIThreadTaskRunner({})->PostTask(FROM_HERE, base::BindOnce(
-    ^(mojo::Remote<{{cpp_namespace}}::{{interface.name}}> remote){
-      remote.reset();
-    }, std::move(_wrappedValue)));
-}
-
-{{ objc_methods_impls(interface) }}
-
-@end
-
-@implementation {{class_prefix}}{{interface.name}}OwnedImpl {
-  std::unique_ptr<{{cpp_namespace}}::{{interface.name}}> _wrappedValue;
-{%- for bridge in interface|interface_remote_sets %}
-  std::vector<base::SequenceBound<std::unique_ptr<{{bridge}}>>> _{{bridge|under_to_lower_camel}}Receivers;
-{%- endfor %}
-}
-
-- (instancetype)initWith{{interface.name}}:(std::unique_ptr<{{cpp_namespace}}::{{interface.name}}>)obj {
+- (instancetype)initWith{{interface.name}}Impl:(std::unique_ptr<{{cpp_namespace}}::{{interface.name}}>)obj {
   if ((self = [super init])) {
-    _wrappedValue = std::move(obj);
+    mojo::PendingReceiver<{{cpp_namespace}}::{{interface.name}}> receiver;
+    _wrappedValue = mojo::Remote<{{cpp_namespace}}::{{interface.name}}>(receiver.InitWithNewPipeAndPassRemote());
+    mojo::MakeSelfOwnedReceiver(std::move(obj), std::move(receiver), web::GetUIThreadTaskRunner({}));
   }
   return self;
 }
 
 - (void)dealloc {
-  web::GetUIThreadTaskRunner({})->DeleteSoon(FROM_HERE, std::move(_wrappedValue));
+  web::GetUIThreadTaskRunner({})->PostTask(FROM_HERE, base::BindOnce(
```

Comment:
Is this needed since the remote is destructed on the UI task already?
It shouldn't be necessary I think.

Reply:
This isn't actually a new change, not sure why git presented it like that. I don't think it's guaranteed anywhere that remote deconstructs off main but if it is then I can remove this

Reply:
> > // If invoked at all, |[handler](https://source.chromium.org/chromium/chromium/src/+/main:mojo/public/cpp/bindings/remote.h;drc=105770df485ace262780d95126bb60b1a16ec340;l=146?q=mojo::Remote&ss=chromium%2Fchromium%2Fsrc)| will be scheduled asynchronously using the
>   // Remote's bound SequencedTaskRunner.

It should automatically be scheduled for deleting on its internal TaskRunner. So this line isn't needed afaict. However, I'll approve anyway as it doesn't hurt anything.

Reply:
Based on what I see in the headers thats specifically on the [set_disconnect_handler](https://source.chromium.org/chromium/chromium/src/+/main:mojo/public/cpp/bindings/remote.h;drc=105770df485ace262780d95126bb60b1a16ec340;bpv=1;bpt=1;l=146?gsn=set_disconnect_handler&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fchromium.googlesource.com%2Fcodesearch%2Fchromium%2Fsrc%2F%2Fmain%3Flang%3Dc%252B%252B%3Fpath%3Dmojo%2Fpublic%2Fcpp%2Fbindings%2Fremote.h%23VQ5Vq3SqSW9Z3rMAtlzEuxOv5eatDcDgB3Nc0MggOWQ), which technically is called if you call `reset_on_disconnect` but I dont think thats guaranteed for all remotes